build-tools/packages/build-cli/src/lib/bump.ts

```diff
@@ -160,14 +174,11 @@
 	let cmd: string;
 	let workingDir: string;
 
+	// Run npm version in each package to set its version in package.json. Also regenerates packageVersion.ts if needed.
 	if (releaseGroupOrPackage instanceof MonoRepo) {
 		workingDir = releaseGroupOrPackage.repoPath;
 		name = releaseGroupOrPackage.kind;
-		cmd = `npx --no-install lerna version ${
-			translatedVersion.version
-		} --no-push --no-git-tag-version -y${
-			exactDependencyType === "" ? " --exact" : ""
-		} && npm run build:genver`;
+		cmd = `flub exec -g ${name} -- "npm version '${translatedVersion.version}'" && npm run build:genver`;
```

Comment:
The input is protected by oclif flag parsing and typing; it's not arbitrary string input.

Reply:
its generally better practice to not involve the shell when trying to programmatically run commands. Relying on it to reparse the arguments as you expect (and not do any shell substitutions, have issues with quoting escapes, unexpected environment variables etc) is a common case of security issues and platform specific errors.

Its much better to invoke a specific executable with a given list of arguments when possible. More secure, simpler to reason about, way fewer edge cases, and faster since you don't need to spawn a shell process just to un-concat your aguments (and check for escapes, and substantiations).

I'm not sure how you do this with node, but I'm sure there is an API for it.

Reply:
For an example of how this could go wrong, in the future if someone breaks the quoting, "~" would get expanded into the user directory of our build machine when publishing the package. Probably not going to happen or horrible if it does, but we can avoid the risk of issues like that if you don't needlessly run strings through the shell before passing them to the actual command

Reply:
Actually, this doesn't run on CI. Which raises the question: is this where we should be doing this? I don't think so.

In repo we should probably use *workspace versions (I've helped people debug merge results that caused them to download packages when they thought they were getting workspace versions: its confusing and unnecessary). If we used workspace versions we wouldn't need this pump logic at all.

We already have logic for adjusting the deps as part of publishing. Prerelease builds for example refer to exact versions with the build numbers: I think it would make more sense to make the checked in versions all "workspace*", not dop bumping, and set the dep type in the publishing logic.

Reply:
We should be using `child_process.spawn(command[, args][, options])` instead of exec unless we really need the shell for something.

Reply:
> In repo we should probably use *workspace versions (I've helped people debug merge results that caused them to download packages when they thought they were getting workspace versions: its confusing and unnecessary). If we used workspace versions we wouldn't need this pump logic at all.

I'm separately working on workspace protocol support. We can't use bump tools as-is because of our internal version scheme.

> We already have logic for adjusting the deps as part of publishing. Prerelease builds for example refer to exact versions with the build numbers: I think it would make more sense to make the checked in versions all "workspace*", not dop bumping, and set the dep type in the publishing logic.

This is already true: https://github.com/microsoft/FluidFramework/blob/2ee97e3c084c68e6d736e9f75ebef250f48f3322/tools/pipelines/templates/include-set-package-version.yml#L113-L140

`flub bump` is used to set the versions/dep ranges for client because of the weird version scheme. I can't remember why I didn't include the pipeline changes in this PR - I'll do that.

> We should be using child_process.spawn(command[, args][, options]) instead of exec unless we really need the shell for something.

I'll switch.

Reply:
> >We should be using child_process.spawn(command[, args][, options]) instead of exec unless we really need the shell for something.
>
>I'll switch.

Changed my mind. Since this change is going to be in `flub exec`, I'm going to do it separately.